File: projects/green_house/picoW/web/bridge_server.ts
import http from 'node:http';
import fs from 'node:fs';
import path from 'node:path';
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';

// Minimal TypeScript Bridge Server
// Run with: node --experimental-strip-types bridge_server.ts (for newer Node versions)
// Or simply: node bridge_server.ts (if Node 23+)

const PORT = 8000;
const __dirname = path.dirname(fileURLToPath(import.meta.url));

const scripts: Record<string, string> = {
    python: "/home/tom/TamiAlQuida.github.io/projects/old_historic/bolan.py",
    bash: "/home/tom/TamiAlQuida.github.io/projects/old_historic/bme280_sensor.sh",
    cpp: "/home/tom/TamiAlQuida.github.io/projects/old_historic/controller.cpp"
};

const MIME_TYPES: Record<string, string> = {
    '.html': 'text/html',
    '.js': 'text/javascript',
    '.css': 'text/css',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
};

const server = http.createServer((req, res) => {
    // Set CORS headers for all responses
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, GET, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    // Handle preflight OPTIONS
    if (req.method === 'OPTIONS') {
        res.writeHead(204);
        res.end();
        return;
    }

    // Handle Script Execution
    if (req.method === 'POST' && req.url === '/run-script') {
        let body = '';
        req.on('data', chunk => { body += chunk; });
        req.on('end', () => {
            try {
                const data = JSON.parse(body);
                const scriptType = data.type as string;

                if (scripts[scriptType]) {
                    const targetPath = scripts[scriptType];
                    console.log(`Executing ${scriptType} script: ${targetPath}`);

                    let execCmd = '';
                    if (scriptType === 'python') {
                        execCmd = `python3 ${targetPath}; echo; echo 'Process finished. Press Enter to close...'; read`;
                    } else if (scriptType === 'bash') {
                        execCmd = `bash ${targetPath}; echo; echo 'Process finished. Press Enter to close...'; read`;
                    } else {
                        execCmd = `echo 'C++ Script info: ${targetPath}'; echo; echo 'Press Enter to close...'; read`;
                    }

                    // Launch Alacritty terminal
                    spawn('alacritty', ['-e', 'bash', '-c', execCmd], {
                        detached: true,
                        stdio: 'ignore'
                    }).unref();

                    res.writeHead(200, { 'Content-Type': 'application/json' });
                    res.end(JSON.stringify({ status: 'success', message: `Started ${scriptType}` }));
                } else {
                    res.writeHead(400, { 'Content-Type': 'application/json' });
                    res.end(JSON.stringify({ status: 'error', message: 'Unknown script type' }));
                }
            } catch (e) {
                res.writeHead(400, { 'Content-Type': 'application/json' });
                res.end(JSON.stringify({ status: 'error', message: 'Invalid JSON' }));
            }
        });
        return;
    }

    // Handle Static Files
    let filePath = req.url === '/' ? '/index.html' : req.url || '';
    const fullPath = path.join(__dirname, filePath);

    const ext = path.extname(fullPath).toLowerCase();
    const contentType = MIME_TYPES[ext] || 'application/octet-stream';

    fs.readFile(fullPath, (error, content) => {
        if (error) {
            if (error.code === 'ENOENT') {
                res.writeHead(404);
                res.end('File Not Found');
            } else {
                res.writeHead(500);
                res.end(`Server Error: ${error.code}`);
            }
        } else {
            res.writeHead(200, { 'Content-Type': contentType });
            res.end(content, 'utf-8');
        }
    });
});

server.listen(PORT, '0.0.0.0', () => {
    console.log('==================================================');
    console.log(`  TYPESCRIPT BRIDGE SERVER RUNNING ON http://localhost:${PORT}`);
    console.log('  1. Serves static files');
    console.log('  2. Runs local scripts via alacritty');
    console.log('==================================================');
});

File: projects/green_house/picoW/web/bridge_server.cpp
#include "httplib.h"
#include <cstdlib>
#include <iostream>
#include <map>
#include <string>
#include <vector>

// Helper to extract a single string value from a simple JSON like {"type":
// "value"}
std::string get_json_value(const std::string &body, const std::string &key) {
  std::string search_key = "\"" + key + "\":";
  size_t pos = body.find(search_key);
  if (pos == std::string::npos)
    return "";

  size_t start = body.find("\"", pos + search_key.length());
  if (start == std::string::npos)
    return "";

  size_t end = body.find("\"", start + 1);
  if (end == std::string::npos)
    return "";

  return body.substr(start + 1, end - start - 1);
}

int main() {
  httplib::Server svr;

  // Configuration
  const int PORT = 8000;
  std::map<std::string, std::string> scripts = {
      {"python",
       "/home/tom/TamiAlQuida.github.io/projects/old_historic/bolan.py"},
      {"bash", "/home/tom/TamiAlQuida.github.io/projects/old_historic/"
               "bme280_sensor.sh"},
      {"cpp",
       "/home/tom/TamiAlQuida.github.io/projects/old_historic/controller.cpp"}};

  // 1. Serve static files from the current directory
  auto ret = svr.set_mount_point("/", "./");
  if (!ret) {
    std::cerr << "Error: Directory not found." << std::endl;
    return 1;
  }

  // 2. Handle POST requests to /run-script
  svr.Post("/run-script", [&](const httplib::Request &req,
                              httplib::Response &res) {
    std::string script_type = get_json_value(req.body, "type");

    // Add CORS headers
    res.set_header("Access-Control-Allow-Origin", "*");

    if (scripts.find(script_type) != scripts.end()) {
      std::string target_path = scripts[script_type];
      std::cout << "Executing " << script_type << " script: " << target_path
                << std::endl;

      std::string exec_cmd;
      if (script_type == "python") {
        exec_cmd =
            "python3 " + target_path +
            "; echo; echo 'Process finished. Press Enter to close...'; read";
      } else if (script_type == "bash") {
        exec_cmd =
            "bash " + target_path +
            "; echo; echo 'Process finished. Press Enter to close...'; read";
      } else {
        exec_cmd = "echo 'C++ Script info: " + target_path +
                   "'; echo; echo 'Press Enter to close...'; read";
      }

      // Launch Alacritty in background
      std::string full_cmd = "alacritty -e bash -c \"" + exec_cmd + "\" &";
      std::system(full_cmd.c_str());

      res.set_content("{\"status\": \"success\", \"message\": \"Started " +
                          script_type + "\"}",
                      "application/json");
    } else {
      res.status = 400;
      res.set_content(
          "{\"status\": \"error\", \"message\": \"Unknown script type\"}",
          "application/json");
    }
  });

  // 3. Handle OPTIONS for CORS (pre-flight)
  svr.Options("/run-script",
              [](const httplib::Request &, httplib::Response &res) {
                res.set_header("Access-Control-Allow-Origin", "*");
                res.set_header("Access-Control-Allow-Methods", "POST, OPTIONS");
                res.set_header("Access-Control-Allow-Headers", "Content-Type");
                res.status = 200;
              });

  std::cout << "=================================================="
            << std::endl;
  std::cout << "  C++ BRIDGE SERVER RUNNING ON http://localhost:" << PORT
            << std::endl;
  std::cout << "  1. Serves static files" << std::endl;
  std::cout << "  2. Runs local scripts via alacritty" << std::endl;
  std::cout << "=================================================="
            << std::endl;

  svr.listen("0.0.0.0", PORT);

  return 0;
}
